pkg/mutation/system.go

```diff
@@ -98,14 +106,72 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (b
 				return false, nil
 			}
 			if cmp.Equal(original, obj) {
+				if *MutationLoggingEnabled {
+					logAppliedMutations("Oscillating mutation.", original, allAppliedMutations)
+				}
 				return false, fmt.Errorf("oscillating mutation for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 			}
+			if *MutationLoggingEnabled {
+				logAppliedMutations("Mutation applied.", original, allAppliedMutations)
+			}
 			return true, nil
 		}
+		if *MutationLoggingEnabled {
+			allAppliedMutations = append(allAppliedMutations, appliedMutations)
+		}
+	}
+	if *MutationLoggingEnabled {
+		logAppliedMutations("Mutation not converging.", original, allAppliedMutations)
 	}
 	return false, fmt.Errorf("mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 }
 
+func logAppliedMutations(message string, obj *unstructured.Unstructured, allAppliedMutations [][]types.Mutator) {
+	iterations := []interface{}{}
+	for i, appliedMustations := range allAppliedMutations {
+		if len(appliedMustations) == 0 {
+			continue
+		}
+		var appliedMutationsText strings.Builder
+		mutationSeparator := ""
+		for _, mutator := range appliedMustations {
+			fmt.Fprintf(&appliedMutationsText, "%s%s", mutationSeparator, mutatorKey(mutator))
+			mutationSeparator = ", "
+		}
+		iterations = append(append(iterations, fmt.Sprintf("iteration_%d", i)), appliedMutationsText.String())
+	}
+	if len(iterations) > 0 {
+		logDetails := []interface{}{}
+		logDetails = append(append(logDetails, logging.EventType), logging.MutationApplied)
+		logDetails = append(append(logDetails, logging.ResourceGroup), obj.GroupVersionKind().Group)
+		logDetails = append(append(logDetails, logging.ResourceKind), obj.GroupVersionKind().Kind)
+		logDetails = append(append(logDetails, logging.ResourceAPIVersion), obj.GroupVersionKind().Version)
+		logDetails = append(append(logDetails, logging.ResourceNamespace), obj.GetNamespace())
+		logDetails = append(append(logDetails, logging.ResourceName), obj.GetName())
+		logDetails = append(logDetails, iterations...)
+		log.Info(message, logDetails...)
+	}
+}
+
+func mutatorKey(mutator types.Mutator) string {
+	mutatorType := ""
+	resourceVersion := ""
+	switch castMutator := mutator.(type) {
+	case *AssignMutator:
+		mutatorType = "Assign"
+		resourceVersion = castMutator.assign.GetResourceVersion()
```

Comment:
Is it possible to enter this case but for castMutator to be nil?

Reply:
Not realy. The mutator can not be nil, and it's either AssignMutator or AssignMetadataMutator. Someone would have to actually add a new kind of mutator to the code and recompile for this to hapen.
I had the code for a switch default for this, but thought that the additional error returns and "if err!=nil" are not worth it.

Reply:
Code removed after latest changes